### ISSUE TITLE
Course search feature

### ISSUE THREAD

**Issue by jaguarfi** [apluslms/a-plus#522]
>**Migrated from Trello**

Update A+ front page with course search tools. This would include fields like:

* text search (code, name, instance name)
* visibility (aalto, fitech, external)
* when (past, current, future)

Search results would be organised in groups future, current and past instances. Books could be own category.

With this, we could create link to other systems like `plus.cs.aalto.fi/?code=cs-a1110` and user would see list of current, future and past instances.

With this, we could create a better user experience than with "current" url. With query like `?code=cs-a1110&when=current&next=/foo/bar&forward=true` user would be automatically forwarded to a current active instance's page `/foo/bar`. If there is no current instance, then system would show list of future and past instances and would show if the next url is valid within those. If the next url is valid in a past course and user clicks the course, they are again forwarded to the correct page.

When linking from a mooc.aalto.fi, we can use filter `?visibility=external` and then only MOOC courses would be listed. Similarly, if linked from fitech page, we could use `?visibility=fitech` to limit courses.

----

**Conversation**

>**Juha Sorva:**
I really like this proposal. As an addition, it would be nice to have a very short and crisp URL to refer to (all the instances of) a course, perhaps using a simple ID instead of official course code, e.g. ``plus.cs.aalto.fi/o1`` and/or ``plus.cs.aalto.fi/cs-a1110``

>**Jaakko Kantojärvi :**
Yeah. We could use the course key to show all instances of that course. Good point.

>This card should fulfil an old idea described here: links of the form #issue

>**Juha Sorva:**
**@raphendyr** Yes, and perhaps support the params there too, as in ``https://plus.cs.aalto.fi/o1?visibility=external``

**Comment by jaguarfi** [comment]
I am not sure about using query parameters for this. At least for the course instances, the approach suggested by @jsorva makes more sense. It is easier and more practical to share and remember a plain URL that a URL with a bunch of query parameters.

>Perhaps there could be one URL for the list of instances (e.g., plus.cs.aalto.fi/o1) and another for the most recent instance (e.g., plus.cs.aalto.fi/o1/current).
>
>Alternatively plus.cs.aalto.fi/o1 for the recent instance and plus.cs.aalto.fi/o1/instances or similar for the list.
https://github.com/apluslms/a-plus/issues/746#issuecomment-754688973

**Comment by Mankro** [comment]
> I am not sure about using query parameters for this. At least for the course instances, the approach suggested by @jsorva makes more sense. It is easier and more practical to share and remember a plain URL that a URL with a bunch of query parameters.
> 
> > Perhaps there could be one URL for the list of instances (e.g., plus.cs.aalto.fi/o1) and another for the most recent instance (e.g., plus.cs.aalto.fi/o1/current).
> > Alternatively plus.cs.aalto.fi/o1 for the recent instance and plus.cs.aalto.fi/o1/instances or similar for the list.
> > [#746 (comment)](https://github.com/apluslms/a-plus/issues/746#issuecomment-754688973)

Now you are mixing issues. Your comment makes more sense in https://github.com/apluslms/a-plus/issues/578, but nobody has suggested using URL query parameters there. This issue is about the search feature: searching for courses based on textual query. Search is needed when you are looking for courses without knowing exactly which course is the right one.

**Comment by jaguarfi** [comment]
Since there is a `when+current` query parameter in the description of this issue I thought that the suggestion was to use the `when=` query parameter to search the current instance.

**Comment by Mankro** [comment]
Yes true, the description mixes ideas from general course search and the links to the latest instance of a course. There are different ideas for how to approach this. We can agree that we don't really want to make the latest course instance URLs very complicated with query parameters, at least not if there is no other easier URL for the latest instance.

I wanted to emphasize that the general search and the link to the latest course instance are different features that are covered in different issues. However, the old description of this issue also mixes different ideas together, which adds confusion. Though, this is also a meta issue that links to sub-issues.

The giveaway here is that there are many different ideas for the implementation and we will design the final version only when we actually work on this, which will probably not be in the near future.

**Comment by Mankro** [comment]
Now we have implemented the list page for course instances and the redirect view for the latest course instance. https://github.com/apluslms/a-plus/pull/772
